pkg/clients/frontegg_client.go

```diff
@@ -0,0 +1,238 @@
+package clients
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+)
+
+// FronteggClient struct to encapsulate the http.Client with additional properties
+type FronteggClient struct {
+	HTTPClient  *http.Client
+	Token       string
+	Email       string
+	Endpoint    string
+	TokenExpiry time.Time
+	Password    string
+}
+
+// NewFronteggClient function for initializing a new Frontegg client with an auth token
+func NewFronteggClient(ctx context.Context, password, endpoint string) (*FronteggClient, error) {
+	token, email, tokenExpiry, err := getToken(ctx, password, endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token: %v", err)
+	}
+
+	transport := &tokenTransport{
+		Token:     token,
+		Transport: http.DefaultTransport,
+	}
+
+	client := &http.Client{Transport: transport}
+
+	return &FronteggClient{
+		HTTPClient:  client,
+		Token:       token,
+		Email:       email,
+		Endpoint:    endpoint,
+		TokenExpiry: tokenExpiry.Add(-time.Duration(0.5*float64(time.Until(tokenExpiry).Nanoseconds())) * time.Nanosecond),
+		Password:    password,
+	}, nil
+}
+
+// tokenTransport struct to add the Authorization header to each request
+type tokenTransport struct {
+	Token     string
+	Transport http.RoundTripper
+}
+
+// RoundTrip method to execute the request with the token
+func (t *tokenTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Deep copy the request to ensure it's safe to modify
+	req2 := cloneRequest(req)
+	req2.Header.Set("Authorization", "Bearer "+t.Token)
```

Comment:
Can you say more about why we need the deep copy?

Reply:
This is primarily related to concurrency and the immutability of the original request. As http.Request.Body can only be read once, a new body needs to be copied:

> https://stackoverflow.com/questions/62017146/http-request-clone-is-not-deep-clone

Will try to adjust the comment to give a bit more details!

Reply:
How about you include the stack overflow link in the comment.

Reply:
Sounds good, just added the link in the func comments.